src/main/kotlin/me/zeroeightsix/kami/feature/module/Notebot.kt

```diff
@@ -0,0 +1,212 @@
+        playButton()
+        pauseButton()
+    }
+
+    private fun playButton() {
+        (playingSong).conditionalWrap(
+            {
+                ImGui.pushItemFlag(ItemFlag.Disabled.i, true)
+                ImGui.pushStyleVar(StyleVar.Alpha, ImGui.style.alpha * 0.5f)
+            },
+            {
+                dsl.button("Play") {
+                    playingSong = true
+                }
+            },
+            {
+                ImGui.popItemFlag()
+                ImGui.popStyleVar()
+            }
+        )
+    }
+    private fun pauseButton() {
+        (!playingSong).conditionalWrap(
+            {
+                ImGui.pushItemFlag(ItemFlag.Disabled.i, true)
+                ImGui.pushStyleVar(StyleVar.Alpha, ImGui.style.alpha * 0.5f)
+            },
+            {
+                dsl.button("Pause") {
+                    playingSong = false
+                }
+            },
+            {
+                ImGui.popItemFlag()
+                ImGui.popStyleVar()
+            }
+        )
+    }
+
+    private fun discoverSong() {
+        mc.player?.let {
+            for (x in -4..4) {
+                for (y in -4..4) {
+                    for (z in -4..4) {
+                        val pos = it.pos?.plus(Vec3d(x.toDouble(), y.toDouble(), z.toDouble())) ?: break
+                        if (it.world.getBlockState(BlockPos(pos))?.block == Blocks.NOTE_BLOCK) {
+                            val noteBlock = it.world.getBlockState(BlockPos(pos))
+                            displayActionbar(it, "" + noteBlock.get(INSTRUMENT) + "[" + noteBlock.get(NOTE) + "]")
+                            instrumentMap.add(noteBlock.get(INSTRUMENT), noteBlock.get(NOTE), BlockPos(pos))
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun loadSong() {
+        mc.player?.let {
+            playingSong = false
+            try {
+                MidiParser.parse("songs${File.separator}$songName").let {
+                    noteSequence = it.first
+                    channelMap = it.second
+                }
+                displayActionbar(it, "Loaded song ${File.separator}$songName")
+                lastNote = System.currentTimeMillis()
+                elapsed = 0
+            } catch (e: Exception) {
+                displayActionbar(it, "Sound not found ${File.separator}$songName")
+            }
+        }
+    }
+
+    @EventHandler
+    val listener = Listener<TickEvent.InGame>({
+        if (playingSong) {
+            val player = it.player
+            val world = it.world
+            if (!player.isCreative) {
+                while (noteSequence.isNotEmpty() && noteSequence.firstKey() <= elapsed) {
+                    playNotes(noteSequence.pollFirstEntry().value, player, world)
+                }
+                if (noteSequence.isEmpty()) playingSong = false
+                elapsed += System.currentTimeMillis() - lastNote
```

Comment:
Instead of keeping track of the elapsed ms cumulatively, remove `lastNote` and opt for storing the `firstNote` time instead:
```suggestion
                elapsed = System.currentTimeMillis() - firstNote
```
You will also be able to remove the line resetting `lastNote` on each tick, and produce more accurate and bug-potential-free code.